### ISSUE TITLE
Filter mapping by ID instead of label

### ISSUE THREAD

**Issue by pawelru** [insightsengineering/teal#969]
Conceptually, I find it quite odd to map filter to module _labels_. It should be a module _identifier_ instead. A label should be a free text but we tend to use it as identifier which is probably not the right thing to do.

**Comment by chlebowa** [comment]
Aren't module ids created from the labels behind the scenes and never made available to the user?

**Comment by pawelru** [comment]
Yes and I don't feel it's correct. Like I said earlier, a label should be a free text, in particular: an empty character or non-unique values between modules. These all should be allowed but it's not because we treat labels as identifier.

**Comment by chlebowa** [comment]
That is a fair point, however, juggling labels AND ids is additional burden on the user.

**Comment by pawelru** [comment]
Yes. We can probably came up with some convenience features but I agree that managing two attributes is more or less twice as difficult as managing just one. 

This doesn't change the point that IMHO right now labels are misused as ids. Recent rollout of mapping feature just exposed such flaw. Until this, the concept of module id was just useless for the user. I think we need to rethink the concept of module ID. Until a new not-yet-anticipated feature / extension that requires an ID.

**Comment by lcd2yyz** [comment]
I remember we discussed this briefly during the design of the module-specific mechanism, but decided to stick with the labels for the sake of moving on, because there weren't clear/better alternatives at the time. I concur @pawelru's thoughts, thanks for raising the topic again.

**Comment by gogonzo** [comment]
We are in dead end with this one.

1. Issue with mapping argument - even if we decide to do following

```r
modules(
  tm_xy("mod1", ,..., filters = teal_slices(teal_slice_age, teal_slice_sex),
  modules(
    label = "tab",
    tm_xy("mod1", ...., filters = teal_slices(teal_slice_age)
  )
)
``` 

We would be able to map modules properly together even if they are nested. 


2. Problem with displaying mapping in the app - here situation is different because we need to have a visual depresentation of these modules. For above example we would display (currently it is not possible because we don't allow duplicated labels when module-specific):

| filter | mod1 | mod 1 |
| ---- | ---- | ---- |
| AGE |  x | x |
| SEX | x | o |


And we rather should display this information in following way:

| filter | mod1 | tab > mod 1 |
| ---- | ---- | ---- |
| AGE |  x | x |
| SEX | x | o |

**Comment by gogonzo** [comment]
Another solution is:

- teal to force `label` change when duplicated.
- remove nested modules

**Comment by pawelru** [comment]
That's exactly what I was referring to. Use `ID` instead of `label` which should remain as a free text that we can further combine or anything.

---

How about the following: add optional (!) `id` argument to the module. We can make a reasonable default value (e.g. `sprintf("%s_%s", short_hash(Sys.time()), label_to_id(label))` OR auto-create it inside the `init()`. Regardless of this, we would also need to add uniqueness check somewhere inside `init()`.

This would address @chlebowa concern about throwing yet another technical requirement to the app developer which is totally valid point.

> Problem with displaying mapping in the app

I believe this would be sorted out as well. We will be having the following structure: 
```
[
  {
    id: some_id1, 
    label: "mod1", 
    filters: [...]
  },
  {
    id: some_id2,
    label: "tab > mod2",
    filters: [...]
  }
]
```

From this, you can easily visualise this as a table and skip ID at all.
